### PR TITLE
Convert ABIs to TypeScript

### LINKED ISSUE #2: [MarketToken] Test converted TypeScript ABI
## Before you `start work`

Please read our contribution [guidelines](https://docs.marketprotocol.io/#contributing) and if there is a bounty involved  please also see [here](https://docs.marketprotocol.io/#gitcoin-and-bounties)

If you have ongoing work from other bounties with us where funding has not been released, please do not pick up a new issue.  We would like to involve as many contributors as possible and parallelize the work flow as much as possible.

Please make sure to comment in the issue here immediately after starting work so we know your plans for implementation and a timeline.

Please also note that in order for work to be accepted, all code must be accompanied by test cases as well.

### User Story

As a developer, I need to ensure the MARKET API TypeScript ABIs work correctly so that I can correctly interface with the MARKET contracts.

### Why Is this Needed?

*Summary*: We need to have automated test coverage for all TypeScript ABIs. This started as part of https://github.com/MARKETProtocol/MARKET.js/issues/14. Due to the size and scope of the testing, we created a new project and repo for the MARKET.js types.

### Description

*Type*: Enhancement

### Current Behavior
There is minimal test coverage.

### Expected Behavior
We need 100% test coverage on the TypeScript ABIs.

### Solution
*Summary*:
[Typechain](https://github.com/Neufund/TypeChain) is being used to generate types and the ABIs from which to generate the types is specified as a glob pattern in the package.json under config.typechain.include which is currently set to: Market*|CollateralToken|OraclizeQueryTest|ERC20|OrderLib

### Definition of Done
- Create tests for the MarketToken ABI

### PR DESCRIPTION
**Pull request by ConnorChristie**
#### Description
This converts the ABIs into their Typescript representatives with unit tests to ensure all methods have been appropriately generated.

#### Implementation details
[Typechain](https://github.com/Neufund/TypeChain) is being used to generate types and the ABIs from which to generate the types is specified as a glob pattern in the package.json under config.typechain.include which is currently set to: `Market*|CollateralToken|OraclizeQueryTest|ERC20|OrderLib`

#### Original issue
https://github.com/MARKETProtocol/MARKET.js/issues/14

#### TODO
- [x] MarketToken tests (resolves #2)
- [ ] CollateralToken tests
- [ ] ERC20 tests
- [ ] MarketCollateralPool tests
- [ ] MarketContract tests
- [ ] MarketContractFactoryOraclize tests
- [ ] MarketContractOraclize tests
- [ ] MarketContractRegistry tests
- [ ] OraclizeQueryTest tests
- [ ] OrderLib tests

There are quite a few tests that have yet to be added which is quite laborious, if anyone is willing to help out, that would be very appreciated!

### CODE REVIEW

**Comment by pelsasser** [comment]
@ConnorChristie Thanks so much for the PR! We can definitely get some more hands on deck to help with all the testing, happy to stake some more issues with gitcoin to get some traction there once this gets merged. 

Will let @eswarasai coordinate that, but @ConnorChristie thanks for all your hard work!

**Comment by eswarasai** [comment]
@ConnorChristie -- Overall, the PR looks great to me expect for few minor tweaks here and there. Please find them below:

- Existing tests have few linting errors. Need to fix them. Please try to add strict typing wherever possible
- I feel it might be better to have a `constants.ts` in the `tests` folder to easily import across multiple test cases. Can you refactor this please?
- Is `web3` and `bignumber.js` needed as `dependency` or can be moved to `devDependency`?
- It makes sense to install `@0xproject/typescript-typings` to have appropriate typings for `web3`
- Can you please upgrade the initial version to `0.1.0` instead of `0.0.1`?
- Can you add [`.travis.yml`](https://github.com/MARKETProtocol/MARKET.js/blob/develop/.travis.yml)? I'll see if I can setup Travis CI for this repo

If adding tests would be too much of a task for you, please do create GitHub issues in this repo so that we can stake them out. Thanks for all your hard work on this :)

**Comment by ConnorChristie** [comment]
- Existing tests have few linting errors. Need to fix them. Please try to add strict typing wherever possible
  - My linter is show these `Warning: The 'await-promise' rule requires type information.` if that's what you're referring to? I have no idea how to fix them... And refactored test code to introduce more strong typings.

- I feel it might be better to have a constants.ts in the tests folder to easily import across multiple test cases. Can you refactor this please?
  - Done

- Is web3 and bignumber.js needed as dependency or can be moved to devDependency?
  - They are needed during runtime by the `typechain-runtime` so yes. I am not sure the effect of moving them to the devDependencies though. I tried setting them as peer deps but travis doesn't like that.

- It makes sense to install @0xproject/typescript-typings to have appropriate typings for web3
  - Done

- Can you please upgrade the initial version to 0.1.0 instead of 0.0.1?
  - Done

- Can you add .travis.yml? I'll see if I can setup Travis CI for this repo
  - Done

**Comment by eswarasai** [comment]
@ConnorChristie -- All in all this looks really amazing. Great job 👍 

I have one minor request though:
- `constants.ts` was supposed to have the constants like `contractAddress`, `owner`, `userAddress` etc so that they can be imported and used acrossed multiple test cases without having to hard-code them everywhere. Sorry I wasn't specific initially. But the refactor of `TestContract` is also helpful. I'd recommend moving this to `TestContract.ts` instead of having it in `constants.ts`

File: test/constants.ts [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+    this.mockWeb3Contract = {
+      address: address,
+      abi: require(`../node_modules/market-solidity/build/contracts/${contractName}.json`)
+    };
+
+    spyOn(this.mockWeb3.eth, 'getCode').and.callFake(
+      (addr: string, cb: (err: Error, code: string) => void) => {
+        expect(addr).toEqual(address);
+        cb(null, contractCode || '0x234');
```
Comment (eswarasai) [inline test/constants.ts]: @ConnorChristie -- So by lint errors I meant such as this one --
`null is not assignable to parameter of type Error`

File: test/constants.ts [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+  }
+
+  /**
+   * Mocks out a web3 method
+   *
+   * @param methodName The method to mock out
+   * @param expectedOrError The expected error or return value
+   * @param expectedArgs The expected arguments
+   */
+  public setupMethodSpy(methodName: keyof T, expectedOrError: Error | any, ...expectedArgs: any[]) {
```
Comment (eswarasai) [inline test/constants.ts]: @ConnorChristie -- Can we hardcode `methodName` of type `string` instead of `keyof T`?

File: test/constants.ts (hunk 1 above)
Comment (ConnorChristie) [reply test/constants.ts]: I am not sure I see that anywhere... This is all that shows in Travis and my local as well:

```
> tslint -t codeFrame 'types/**/*.ts' 'test/**/*.ts'
Warning: The 'await-promise' rule requires type information.
Warning: The 'no-unused-variable' rule requires type information.
Warning: The 'no-use-before-declare' rule requires type information.
Warning: The 'return-undefined' rule requires type information.
Warning: The 'no-floating-promises' rule requires type information.
Warning: The 'no-unnecessary-qualifier' rule requires type information.
Warning: The 'no-unnecessary-type-assertion' rule requires type information.
Warning: The 'strict-type-predicates' rule requires type information.
```

Comment (eswarasai) [reply test/constants.ts]: @ConnorChristie -- Ahh. That was in the IDE actually. I use IntelliJ Idea. It isn't major, but maybe you can try configuring your IDE to use the `tslint.json` and `tsconfig.json` and confirm you're seeing the similar warnings on your end as well.

**Comment by ConnorChristie** [comment]
@eswarasai I just opened the project up in IntelliJ and those linter errors indeed showed up... Strange as neither the command line or TravisCI caught them. Anyways, I believe I updated all the spots that you requested :)

**Pull request merged**